### PR TITLE
 Fixed #35656 -- Added autodetector_class attribute to migration commands.

### PR DESCRIPTION
**Pull request by LeOndaz**
# Trac ticket number

ticket-35656

# Branch description
Allow changing autodetector class for migration commands

# Checklist
- [x] This PR targets the `main` branch. 
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [ ] I have added or updated relevant tests.
- [ ] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

### CODE REVIEW

**Comment by LeOndaz** [comment]
@sarahboyce sorry I mistakenly thought that deleting the fork won't affect the active PRs, and I seem to have missed up badly, this is the same PR.

I wanted a new fork because I did miss up a few things and I considered re-forking a safer solution in my case

**Comment by LeOndaz** [comment]
@sarahboyce I don't have permission to view the results of the CI

**Comment by sarahboyce** [comment]
> @sarahboyce I don't have permission to view the results of the CI

It won't be an access issue but sometimes the results are lost. I have squashed the commits which has re-triggered the CI, so :crossed_fingers:

**Comment by LeOndaz** [comment]
> > @sarahboyce I don't have permission to view the results of the CI
> 
> It won't be an access issue but sometimes the results are lost. I have squashed the commits which has re-triggered the CI, so :crossed_fingers: 

what's the news 👀

**Comment by sarahboyce** [comment]
> what's the news 👀

I could see it but it's gone again. You can push up to retriever
But maybe worth looking into django-docker-box and running the tests on postgres locally for example
If I remember correctly, the new tests failed as a migration was created (created the table)

**Comment by LeOndaz** [comment]
@sarahboyce figured out the issue but can't resolve it 

It's not detecting the model now, do you have any thoughts about this?

```python
class ModelForAutodetection(models.Model):
    class Meta:
        # Disable auto loading of this model as we load it on our own
        apps = Apps()
```

```python
class CustomMigrationCommandTests(MigrationTestBase):
    @override_settings(MIGRATION_MODULES={"migrations": "migrations.test_migrations"})
    def test_custom_makemigrations_command(self):
        class CustomAutodetector(MigrationAutodetector):
            def changes(self, *args, **kwargs):
                return []

        class CustomMakeMigrationsCommand(MakeMigrationsCommand):
            autodetector_class = CustomAutodetector

        self.assertTableNotExists("migrations_modelforautodetection")
        apps.register_model("migrations", ModelForAutodetection)

        out = io.StringIO()
        command = CustomMakeMigrationsCommand(stdout=out)

        with self.temporary_migration_module() as migration_dir:
            call_command(command, "migrations", verbosity=0)

            init_file = os.path.join(migration_dir, "__init__.py")
            self.assertFalse(os.path.exists(init_file))

            initial_file = os.path.join(migration_dir, "0001_initial.py")
            self.assertFalse(os.path.exists(initial_file))
            self.assertTableExists("model_for_autodetection")
 
    @override_settings(MIGRATION_MODULES={"migrations": "migrations.test_migrations"})
    def test_custom_migrate_command(self):
        class CustomAutodetector(MigrationAutodetector):
            def changes(self, *args, **kwargs):
                return []

        class CustomMigrateCommand(MigrateCommand):
            autodetector_class = CustomAutodetector

        self.assertTableNotExists("migrations_author")

        out = io.StringIO()
        command = CustomMigrateCommand(stdout=out)

        with self.temporary_migration_module():
            self.assertRaises(
                ModuleNotFoundError, call_command, command, "migrations", verbosity=0
            )

        self.assertTableNotExists("migrations_author")
```

```bash
............F.....................................................................................................................................
======================================================================
FAIL: test_custom_makemigrations_command (migrations.test_commands.CustomMigrationCommandTests.test_custom_makemigrations_command)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/unittest/case.py", line 58, in testPartExecutor
    yield
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/unittest/case.py", line 636, in run
    self._callTestMethod(testMethod)
    ^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/unittest/case.py", line 589, in _callTestMethod
    if method() is not None:
    ^^^^^^^^^^^^^^^^^
  File "/Users/leondaz/PycharmProjects/django-contrib/django/django/test/utils.py", line 446, in inner
    return func(*args, **kwargs)
    ^^^^^^^^^^^^^^^^^
  File "/Users/leondaz/PycharmProjects/django-contrib/django/tests/migrations/test_commands.py", line 3330, in test_custom_makemigrations_command
    self.assertTableExists("model_for_autodetection")
    ^^^^^^^^^^^^^^^^^
  File "/Users/leondaz/PycharmProjects/django-contrib/django/tests/migrations/test_base.py", line 39, in assertTableExists
    self.assertIn(table, connections[using].introspection.table_names(cursor))
    ^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/unittest/case.py", line 1154, in assertIn
    self.fail(self._formatMessage(msg, standardMsg))
    ^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/unittest/case.py", line 717, in fail
    raise self.failureException(msg)
    ^^^^^^^^^^^^^^^^^
AssertionError: 'model_for_autodetection' not found in ['auth_group', 'auth_group_permissions', 'auth_permission', 'auth_user', 'auth_user_groups', 'auth_user_user_permissions', 'django_admin_log', 'django_content_type', 'django_migrations', 'django_session', 'django_site', 'migrations_modelwithcustombase', 'migrations_unmigratedmodel']
```

**Comment by sarahboyce** [comment]
@LeOndaz I have pushed up a new version of the tests that I'm hoping will work and reduced the PR to minimal changes

As [previously mentioned](https://github.com/django/django/pull/18445#issuecomment-2322005295), I still think we should have a system check. I think that's the next step :+1:

**Comment by LeOndaz** [comment]
@sarahboyce check this out

**Comment by LeOndaz** [comment]
(bump)

**Comment by sarahboyce** [comment]
> (bump)

I've been at DjangoCon US and now I'm on holiday a week but it's on my list for when I get back 👍

**Comment by LeOndaz** [comment]
> > (bump)
> 
> I've been at DjangoCon US and now I'm on holiday a week but it's on my list for when I get back 👍 

that's a huge gathering of nerds, hope it went well

File: django/core/checks/migrations.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+            Error(
+                "Migrate and makemigrations don't share the same autodetector class. "
+                "Currently, this behavior is not supported.",
+                hint=(
+                    "makemigrations.Command.autodetector_class is {}, but "
+                    "migrate.Command.autodetector_class is {}."
+                ).format(
+                    make_migrations_autodetector.__name__, migrate_autodetector.__name__
+                ),
+                id="migrations.E001",
```
Comment (sarahboyce) [inline django/core/checks/migrations.py:37]: We need to document this check in `docs/ref/checks.txt`

**Review by sarahboyce — COMMENTED** [review]
Small comments, we're nearly there :+1:

**Comment by sarahboyce** [comment]
Can you also squash to a single commit? :1st_place_medal:

**Comment by LeOndaz** [comment]
@sarahboyce ok I will do it, but shouldn't there be an option that squash-merges the PR into a single commit? this is usually the case with all my contributions, do we have any rules about this so that I can get more familiar with them?

**Comment by sarahboyce** [comment]
> @sarahboyce I did squash the commits, the error doesn't seem like it came from the docs I modified, can you check?

I agree this is unrelated

**Comment by LeOndaz** [comment]
@sarahboyce thanks, alert me if anything is needed, I will start another PR for another feature as well, is it fine if I mention you there at that point?

**Comment by felixxm** [comment]
> @sarahboyce thanks, alert me if anything is needed, I will start another PR for another feature as well, is it fine if I mention you there at that point?

There is no need to ping fellows directly on PRs. Leaving a "bump" comments is also not very helpful. If the ticket that you're working on is [in the review queue](https://code.djangoproject.com/query?status=!closed&needs_better_patch=0&needs_tests=0&needs_docs=0&has_patch=1&stage=Accepted&desc=1&order=changetime) there is nothing more you need to do, just wait patiently for the review. It's completely normal that you have to wait a few weeks (or even months) for the review.

**Review by sarahboyce — APPROVED** [review]
Thank you @LeOndaz - welcome onboard :boat:

**Pull request merged**